### PR TITLE
Added QuandlReader (plus tests and doc)

### PR DESCRIPTION
**Pull request by gliptak**
Completing https://github.com/pydata/pandas-datareader/pull/346

Closing https://github.com/pydata/pandas-datareader/pull/360

### CODE REVIEW

File: docs/source/remote_data.rst [hunk 1]
```diff
@@ -79,7 +79,7 @@ Historical dividends from Yahoo! Finance.
 .. _remote_data.yahoo_quotes:
```
Comment (jreback) [inline docs/source/remote_data.rst]: this got changed to 100755 for some reason, can you switch back

something like

``git update-index --chmod -x <file>`` should do it

File: docs/source/remote_data.rst [hunk 2]
```diff
@@ -199,6 +199,32 @@ Available expiry dates can be accessed from the ``expiry_dates`` property.
...
+======
+
+.. ipython:: python
+
+    import pandas_datareader.data as web
+    symbol = 'WIKI/AAPL'  # or 'AAPL.US'
+    df = web.DataReader(symbol, 'quandl', "2015-01-01", "2015-01-05")
+    df.ix['2015-01-04']
+
+.. _remote_data.quandl:
```
Comment (jreback) [inline docs/source/remote_data.rst]: put this right before the Quandl (and move the enigma one right before the Enigma)

File: docs/source/remote_data.rst [hunk 3]
```diff
@@ -199,6 +199,32 @@ Available expiry dates can be accessed from the ``expiry_dates`` property.
...
 
+Quandl
+======
+
+.. ipython:: python
+
+    import pandas_datareader.data as web
+    symbol = 'WIKI/AAPL'  # or 'AAPL.US'
+    df = web.DataReader(symbol, 'quandl', "2015-01-01", "2015-01-05")
+    df.ix['2015-01-04']
```
Comment (jreback) [inline docs/source/remote_data.rst]: use ``df.loc[...]``

File: pandas_datareader/data.py [hunk 4]
```diff
@@ -22,6 +22,7 @@
 from pandas_datareader.enigma import EnigmaReader
 from pandas_datareader.oanda import get_oanda_currency_historical_rates
 from pandas_datareader.nasdaq_trader import get_nasdaq_symbols
```
Comment (jreback) [inline pandas_datareader/data.py]: hmm lots of files look like the got the 100644 -> 100755

File: pandas_datareader/quandl.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
+import re
+
+from pandas_datareader.base import _DailyBaseReader
+
+
+class QuandlReader(_DailyBaseReader):
+
+    """
```
Comment (jreback) [inline pandas_datareader/quandl.py]: can you add a versionadded 0.5.0

File: pandas_datareader/quandl.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        Possible formats:
+        1. DB/SYM: The Quandl 'codes': DB is the database name,
+        SYM is a ticker-symbol-like Quandl abbreviation
+        for a particular security.
+        2. SYM.CC: SYM is the same symbol and CC is an ISO country code,
+        will try to map to the best single Quandl database for that country.
+        Beware of ambiguous symbols (different securities per country)!
+        Note: Cannot use more than a single string because of the inflexible
+        way the URL is composed of url and _get_params in the superclass
+    start : string,
```
Comment (jreback) [inline pandas_datareader/quandl.py]: extra comman

File: pandas_datareader/quandl.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+class QuandlReader(_DailyBaseReader):
+
+    """
+    Returns DataFrame of historical stock prices from symbol, over date
+    range, start to end.
+
+    Parameters
+    ----------
+    symbols : string.
```
Comment (jreback) [inline pandas_datareader/quandl.py]: no period at the end

File: pandas_datareader/quandl.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+            datasetname = 'WIKI'  # default; symbol stays itself
+        elif mm.group(3) == "/":
+            # --- normal Quandl DB/SYM convention:
+            symbol = mm.group(4)
+            datasetname = mm.group(1)
+        elif mm.group(3) == ".":
+            # secondary convention SYM.CountryCode:
+            symbol = mm.group(1)
+            datasetname = self._db_from_countrycode(mm.group(4))
+        base_url = "https://www.quandl.com/api/v3/datasets/"
```
Comment (jreback) [inline pandas_datareader/quandl.py]: make the base_url a class property

File: pandas_datareader/quandl.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        }
+        paramstring = '&'.join(['%s=%s' % (k, v) for k, v in params.items()])
+        return '%s%s/%s.csv?%s' % (base_url, datasetname, symbol, paramstring)
+
+    def _fullmatch(self, regex, string, flags=0):
+        """Emulate python-3.4 re.fullmatch()."""
+        return re.match("(?:" + regex + r")\Z", string, flags=flags)
+
+    def _db_from_countrycode(self, code):
+        map = dict(
```
Comment (jreback) [inline pandas_datareader/quandl.py]: maybe make a class property

File: docs/source/remote_data.rst (hunk 1 above)
Comment (gliptak) [inline docs/source/remote_data.rst]: done

File: docs/source/remote_data.rst (hunk 3 above)
Comment (gliptak) [inline docs/source/remote_data.rst]: done

File: docs/source/remote_data.rst (hunk 2 above)
Comment (gliptak) [inline docs/source/remote_data.rst]: done

File: pandas_datareader/data.py (hunk 4 above)
Comment (gliptak) [inline pandas_datareader/data.py]: corrected

File: pandas_datareader/quandl.py (hunk 5 above)
Comment (gliptak) [inline pandas_datareader/quandl.py]: done

File: pandas_datareader/quandl.py (hunk 7 above)
Comment (gliptak) [inline pandas_datareader/quandl.py]: done

File: pandas_datareader/quandl.py (hunk 6 above)
Comment (gliptak) [inline pandas_datareader/quandl.py]: done

File: pandas_datareader/quandl.py (hunk 8 above)
Comment (gliptak) [inline pandas_datareader/quandl.py]: modeled after Google options _OPTIONS_BASE_URL

File: pandas_datareader/quandl.py (hunk 9 above)
Comment (gliptak) [inline pandas_datareader/quandl.py]: changed

File: docs/source/remote_data.rst [hunk 10]
```diff
@@ -199,6 +199,32 @@ Available expiry dates can be accessed from the ``expiry_dates`` property.
 
 .. _remote_data.enigma:
 
+Quandl
```
Comment (jreback) [inline docs/source/remote_data.rst]: the quandl tag should be *before* the ``Quandle\n------\n``; and move the enigma one *before* ``Enigma``.

File: docs/source/remote_data.rst [hunk 11]
```diff
@@ -199,6 +199,32 @@ Available expiry dates can be accessed from the ``expiry_dates`` property.
...
+======
+
+.. _remote_data.quandl:
+
+Daily financial data (prices of stocks, ETFs etc.) from quandl.com.
```
Comment (jreback) [inline docs/source/remote_data.rst]: add the http link to ``quandl.com``

File: docs/source/remote_data.rst [hunk 12]
```diff
@@ -199,6 +199,32 @@ Available expiry dates can be accessed from the ``expiry_dates`` property.
...
+The symbol names consist of two parts: DB name and symbol name.
+DB names can be all the free ones listed on the Quandl website.
```
Comment (jreback) [inline docs/source/remote_data.rst]: can you add a link here to the 'free ones'

File: docs/source/remote_data.rst [hunk 13]
```diff
@@ -199,6 +199,32 @@ Available expiry dates can be accessed from the ``expiry_dates`` property.
...
+Symbol names vary with DB name; for WIKI (US stocks), they are the common
+ticker symbols, in some other cases (such as FSE) they can be a bit strange.
+Some sources are also mapped to suitable ISO country codes in the dot suffix
+style shown above,
+currently available for BE, CN, DE, FR, IN, JP, NL, PT, UK, US.
```
Comment (jreback) [inline docs/source/remote_data.rst]: if available can you provide a link to the quandl docs about this

File: docs/source/remote_data.rst [hunk 14]
```diff
@@ -199,6 +199,32 @@ Available expiry dates can be accessed from the ``expiry_dates`` property.
...
+
+Unfortunately (as of June 2017), each DB has a different data schema,
```
Comment (jreback) [inline docs/source/remote_data.rst]: I would remove this comment.

File: docs/source/remote_data.rst (hunk 1 above)
Comment (jreback) [inline docs/source/remote_data.rst]: there is also a TOC in this file (its done manually but that's ok). add an entry there as well.

File: docs/source/remote_data.rst (hunk 10 above)
Comment (gliptak) [inline docs/source/remote_data.rst]: moved

File: docs/source/remote_data.rst (hunk 11 above)
Comment (gliptak) [inline docs/source/remote_data.rst]: added

File: docs/source/remote_data.rst (hunk 12 above)
Comment (gliptak) [inline docs/source/remote_data.rst]: added

File: docs/source/remote_data.rst (hunk 13 above)
Comment (gliptak) [inline docs/source/remote_data.rst]: added link to all datasets

File: docs/source/remote_data.rst (hunk 14 above)
Comment (gliptak) [inline docs/source/remote_data.rst]: removed unfortunately

File: docs/source/remote_data.rst (hunk 1 above)
Comment (gliptak) [inline docs/source/remote_data.rst]: updated TOC

**Comment by gliptak** [comment]
@jreback Green after updates

**Comment by jreback** [comment]
thanks @gliptak !

**Pull request merged**